### ISSUE TITLE
System.Numerics.Vectors and Span<T>

### ISSUE THREAD

**Issue by dangi12012** [dotnet/corefx#28433]
Vector3.CopyTo(Span<T> span)

**Comment by ahsonkhan** [comment]
Thanks for the suggestion @dangi12012. Can you please the API review process guidelines [here](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md).

> The issue description should contain a speclet that represents a sketch of the new APIs, including samples on how the APIs are being used. The goal isn't to get a complete API list, but a good handle on how the new APIs would roughly look like and in what scenarios they are being used. Here is [a good example](https://github.com/dotnet/corefx/issues/271).

For reference. here are the existing CopyTo methods on Vector3:
```C#
    public struct Vector3 : IEquatable<Vector3>, IFormattable
    {
        public void CopyTo(float[] array);
        public void CopyTo(float[] array, int index);
    }
```
cc @eerhardt

**Comment by dangi12012** [comment]
`static void Showcase()
        {
            Vector3 v = (3, 2, 1); //Asume we have vector or vector<T> SIMD

            Span<float> span = stackalloc float[3]; //Now we want to implement our own operator not implemented in SIMD
            v.CopyTo(span); //getting single element is very slow for vector<T>. Creating an array is also VERY slow compared to c++ SIMD. Solution: SPAN<T>

            //Take modulus which is not supported by Vectorized math and slow without stackalloc array
            span[0] = span[0] % 2.0f;
            span[1] = span[1] % 2.0f;
            span[2] = span[2] % 2.0f;

            Vector3 result = new Vector3(span); //Boom fast modulus with 0 copy
        }`

**Comment by msedi** [comment]
I would widen this request to initialize a Vector<T> with Span<T>. Currently you always have to allocate an array to feed the new instance of Vector<T> with new data.

In some scenarios where you really need performance (e.g. bilinear interpolation) you need to fill a lot of interpolation partners. Currently, first you have to load your interpolation partners in a newly allocated arrays, which you only need to create your Vector<T>.

For this situations it makes much more sense to stackalloc the memory for your interpolation partners that you refill with new data in a loop. If the stackalloc'ed memory can be Span<T>ned there's no additional cost in allocating memory.

**Comment by tannergooding** [comment]
`Vector<T>` already has `CopyTo(System.Span<T>)`. This was added in https://github.com/dotnet/coreclr/pull/23333.

`Vector2/3/4` do not have it and I would be fine with adding them, but would like the existing proposal (or a new one given it hasn't been updated by the OP in over a year), following the `good example` @ahsonkhan listed https://github.com/dotnet/corefx/issues/28433#issuecomment-375849157 put up before I mark this as `api-ready-for-review`.

**Comment by tannergooding** [comment]
Tuple deconstruction should be in its own issue and follow the good example as well @dangi12012.

**Comment by msedi** [comment]
@tannergooding: You are right, CopyTo is there, but correct me if I'm wrong there's nothing for the constructor, right? So I cannot do something like

```
Span<float> x = stackalloc float[]{10,20,30,40};
Vector<float> v = new Vector<float>(x);
```

Currently only this is possible:
```
float[] x = new float[]{10,20,30,40};
Vector<float> v = new Vector<float>(x);
```

which could cause pressure on the GC when doing this in a loop in x/x direction over many data points.

**Comment by tannergooding** [comment]
The `Vector(Span<T>)` constructor has existed for over a year now and was added in https://github.com/dotnet/coreclr/pull/16733.

The `Vector(ReadOnlySpan<T>)` constructor was added in the https://github.com/dotnet/coreclr/pull/23333 PR I linked above.

**Comment by msedi** [comment]
@tannergooding: Thanks a lot for pointing this out. I did it today, but I looked it up again. According to the docs it only works for .NET Standard 2.1 Preview, whereas I am on .NET Standard 2.0.

**Comment by benaadams** [comment]
But you can do
```csharp
float* x = stackalloc float[] { 10, 20, 30, 40 };
Vector<float> v = Unsafe.ReadUnaligned<Vector<float>>(x);
```

**Comment by msedi** [comment]
@benaadams 👍 This might help. I'll give it a try.
